Create GovStack_Registration_BB_API_application-0.0.1-resolved.json

ingmarvali opened the pull request:
Draft version created. Reference base used from Ukrainian stack (Data factory).


## Description
Describes API endpoints to submit applications to Registration BB and to manage tasks related to the applications. 

## Related Issue
https://govstack-global.atlassian.net/browse/REG-69


## Motivation and Context
Request was made by Govstack technical commitee.

ingmarvali committed:
Create GovStack_Registration_BB_API_application-0.0.1-resolved.json

Draft version created. Reference base used from Ukrainian stack (Data factory).

ingmarvali committed:
Update GovStack_Registration_BB_API_application-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API_application-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API_application-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API_application-0.0.1-resolved.json

First version ready for commenting

ingmarvali committed:
Create GovStack_Registration_BB_API- registration-0.0.1-resolved.json

API description base is taken from Ukrainian Data Factory

ingmarvali committed:
Rename GovStack_Registration_BB_API_template-1.1.0-resolved.json to GovStack_Registration_BB_API_statistics-1.0.0-resolved.json

ingmarvali committed:
Rename GovStack_Registration_BB_API- registration-0.0.1-resolved.json to GovStack_Registration_BB_API-registration-0.0.1-resolved.json

ingmarvali committed:
Rename GovStack_Registration_BB_API_statistics-1.0.0-resolved.json to GovStack_Registration_BB_API_template-1.1.0-resolved.json

ingmarvali committed:
Update and rename GovStack_Registration_BB_API_application-0.0.1-resolved.json to GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API-registration-0.0.1-resolved.json

ingmarvali committed:
Update readme.txt

sleidig commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+		},
+		"version": "0.0.1"
+	},
+	"servers": [{
+		"url": "http://ss2.egovstack.net/r1/eGovStack/COM/11222456/RegistrationBB/creg/",
+		"description": "Information Mediator BB Security server"
+	}],
```
Does it make sense for the generic GovStack BB spec to point to a specific server url? Or is this required for the sandbox?

And what is exactly hosted at this URL?

ingmarvali replied:
I would say this is an example, could be a dummy. 
In general yes, this Server URL is needed for Information Mediator, this is URL for security server of Information Mediator of Sandbox (or any other instance). When the IM sandbox is used then the OpenAPI description should show where the service can be used.    
https://joinup.ec.europa.eu/sites/default/files/event/attachment/2020-10/2020-09-25-How_Petteri.pdf

sleidig commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+					},
+					"code": {
+						"type": "string",
+						"description": "5000"
```
```suggestion
						"description": "500"
```
this should probably be HTTP Status code 500?

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

Reordered the APIs

sleidig commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+					},
+					"processDefinitionId": {
+						"type": "string",
+						"description": "81c4445c-bff6-11ed-afa1-0242ac120002"
```
```suggestion
						"description": "uuid of the process definition of which this is an instance of; e.g. 81c4445c-bff6-11ed-afa1-0242ac120002"
```

sleidig commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+				"description": "Creates application process instance",
+				"operationId": "createProcessInstanceWithForm",
+				"parameters": [{
+						"$ref": "#/components/parameters/Information-Mediator-Client"
```
What's the role of the information mediator here?

ingmarvali replied:
We need the input.

sleidig commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+			"post": {
+				"tags": ["application-registration"],
+				"summary": "Start application process instance with form",
+				"description": "Creates application process instance",
```
```suggestion
				"description": "Creates application process instance for an individual client's request who can do the registration process through the created form instance.",
```
Is that its purpose? (I am still trying to wrap my head around it overall)

ingmarvali replied:
I think it is like this: 
Process- Service that is created to capture users information. 
Instance- Application file created in the service.

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

Descriptions added

sleidig commented on api/GovStack_Registration_BB_API_workflow-0.0.1-resolved.json:
```diff
@@ -0,0 +1,525 @@
+	"openapi": "3.0.0",
+	"info": {
+		"title": "Registration BB API descriptions",
+		"description": "API from Registration BB",
```
What's the difference between the two OpenAPI files? Can we show this in the title and description briefly?

ingmarvali replied:
Good idea.

ingmarvali replied:
I think we should have two files: 
So if we would propose, for a citizen user:
 - get all files of a user 
 - get all available services for a user - this would include a form schema ID needed for each service
 - get all form schemas/ get a specific form schema
 - submit a new file : this will require a dynamic data object that should match the needed form schema

And for Operator users:
 - get a list of services 
 - get the details of a service, including the flow (part B roles) - this would also include  a form schema ID needed for each role to be submitted + available 
 - get the list of files for a specific part B role
 - change the status of a file in a part B role :  this will require a dynamic data object that should match the needed form schema

sleidig commented on api/GovStack_Registration_BB_API_workflow-0.0.1-resolved.json:
```diff
@@ -0,0 +1,525 @@
+					"registered": {
+						"type": "string",
+						"format": "date-time",
+						"example": "2000-10-23T00:00:00.000Z"
```
can we add description explaining difference of "created" and "registered"? Sounds very similar to me.

sleidig commented on api/GovStack_Registration_BB_API_workflow-0.0.1-resolved.json:
```diff
@@ -0,0 +1,525 @@
+					},
+					"assignee": {
+						"type": "string",
+						"example": "42962de0-bdb2-11ed-9397-0242ac120004"
```
if this is an id, what data / endpoint / registry does it point to?

sleidig replied:
similar question for "taskDefinitionKey" actually. Where are those definitions defined? And what's the difference between task definiton and process definition further below?

ingmarvali replied:
I was thinking that this is an operator who is processing the applicant  file. we need to make the operator APIs. 
API for operator users:
 - get a list of services 
 - get the details of a service, including the flow (part B roles) - this would also include  a form schema ID needed for each role to be submitted + available 
 - get the list of files for a specific part B role
 - change the status of a file in a part B role :  this will require a dynamic data object that should match the needed form schema

sleidig commented on api/GovStack_Registration_BB_API_workflow-0.0.1-resolved.json:
```diff
@@ -0,0 +1,525 @@
+						"type": "string",
+						"example": "42962de0-bdb2-11ed-9397-0242ac120004"
+					},
+					"suspended": {
```
what does "suspended" mean/imply for the system?

ingmarvali replied:
Residue, Will remove.

sleidig commented on api/GovStack_Registration_BB_API_workflow-0.0.1-resolved.json:
```diff
@@ -0,0 +1,525 @@
+						"type": "boolean",
+						"example": "true"
+					},
+					"businessKey": {
```
what does businessKey refer to?

ingmarvali replied:
I think it is the same as KEY. we should align it.

sleidig reviewed (commented):
Are the task and process start endpoints the only critical APIs? Or should it also be possible to complete a registration completely via API or the like? (or is that what calls to create and sign would logically do)?

ingmarvali committed:
Update api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json

Co-authored-by: Sebastian <sebastian.leidig@gmail.com>

ingmarvali committed:
Update api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json

Co-authored-by: Sebastian <sebastian.leidig@gmail.com>

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

ingmarvali commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+
+					{
+						"name": "key",
+						"description": "I think this key is issued by the creator app and = id, ",
```
yes, we need to see if this is required. todo

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

ingmarvali commented on api/GovStack_Registration_BB_API-registration-0.0.1-resolved.json:
```diff
@@ -0,0 +1,163 @@
+
+					{
+						"name": "key",
+						"description": "I think this key is issued by the creator app and = id, ",
```
I think it should be businessKey

ingmarvali committed:
Create ExamplePayload.json

ingmarvali committed:
Update and rename GovStack_Registration_BB_API-registration-0.0.1-resolved.json to GovStack_Registration_BB_API-applicant.json

Fail name change

ingmarvali committed:
Update readme.txt

ingmarvali committed:
Update GovStack_Registration_BB_API-applicant.json

Adding new definitions: 
service-key =  ProcessDefinitionKey = Service unique ID
service-version-id = ProcessDefinitionId = service version  ID
application-file-id = ProcessInstanceId - Apllication file id

ingmarvali committed:
Update GovStack_Registration_BB_API-applicant.json

New API added- application file list.

ingmarvali committed:
Update GovStack_Registration_BB_API-applicant.json

New service added

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow-0.0.1-resolved.json

Document upload service

ingmarvali committed:
Update and rename GovStack_Registration_BB_API_workflow-0.0.1-resolved.json to GovStack_Registration_BB_API_operator.json

Change in file name. This is now operator API file.

ingmarvali committed:
Rename GovStack_Registration_BB_API-applicant.json to GovStack_Registration_BB_API_applicant.json

Change in file name

ingmarvali committed:
Update readme.txt

Added a member

ingmarvali committed:
Update readme.txt

Change in definitions

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

ingmarvali committed:
Update GovStack_Registration_BB_API_operator.json

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

ingmarvali committed:
Update GovStack_Registration_BB_API_operator.json

comments

ingmarvali committed:
Create GovStack_Registration_BB_API_workflow.json

Copied from Govstack Workflow BB API specification. Modified to be usable for Registration BB.

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow.json

small updates

ingmarvali committed:
Update and rename GovStack_Registration_BB_API_template-1.1.0-resolved.json to GovStack_Registration_BB_API_statistics.json

Changed file name

ingmarvali committed:
Create GovStack_Registration_BB_API_Schemas.json

Created  schema services

ingmarvali committed:
Rename GovStack_Registration_BB_API_Schemas.json to GovStack_Registration_BB_API_schemas.json

Fail name change

ingmarvali committed:
Create ApplicantSequenceDiagram.mermaid

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow.json

definition change

ingmarvali committed:
Update GovStack_Registration_BB_API_workflow.json

version number example changed

ingmarvali committed:
Update GovStack_Registration_BB_API_schemas.json

Change in API name

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

updates

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

New version of the submit API

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

File upload service v1

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

upload service fix

ingmarvali committed:
Create GovStack_Registration_BB_API_files.json

Split

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

File upload service simplification

ingmarvali committed:
Update GovStack_Registration_BB_API_files.json

additional requirement url property added

ingmarvali committed:
Update GovStack_Registration_BB_API_files.json

small changes

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

refactoring

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

merged services

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

small corrections

ingmarvali committed:
Delete GovStack_Registration_BB_API_files.json

merged

ingmarvali committed:
Update readme.txt

minor changes

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

path change

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

updates

ingmarvali committed:
Create OperatorSequenceDiagram.mermaid

created

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

New service for operators PUT.

ingmarvali committed:
Update OperatorSequenceDiagram.mermaid

additional services added

ingmarvali committed:
Update OperatorSequenceDiagram.mermaid

small fix

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

Operators servces added

ingmarvali committed:
Delete GovStack_Registration_BB_API_operator.json

merged with applicant file

sleidig commented on api/ExamplePayload.json:
For which API endpoint payload is this an example? Would this be easier to understand if it is part of one of the OpenApi spec files directly? Or otherwise maybe further clarify through the file name?

ingmarvali replied:
Any API response or input that has following element:  formData. I think this json is the content of formData. Tell me if any objection. 

E.g. /api/service/{serviceKey}/startWithForm

sleidig commented on api/readme.txt:
```diff
@@ -1,15 +1,21 @@
+serviceKey =  ProcessDefinitionKey = Service unique ID
+applicationFileId = ProcessInstanceId - Apllication file id
+eFormId = form schema
+Applicant= a user who submits data to Registration BB. 
```
```suggestion
**Service** = specific registration process (there can be multiple different services for different registration processes/forms)
serviceKey =  ProcessDefinitionKey = Service unique ID
Applicant = a user who submits data to Registration BB. 
**Application** = individual applicant's file, filling the forms of a service
applicationFileId = ProcessInstanceId - Apllication file id
eFormId = form schema
```

sleidig commented on api/GovStack_Registration_BB_API_applicant.json:
```diff
@@ -0,0 +1,982 @@
+					"eFormId": {
+						"type": "string",
+						"example": "42962de0-bdb2-11ed-9397-0242ac120004",
+						"description": "Form ID used in the data capturing"
```
can we explain the eFormId more clearly. I am not sure how this is different from serviceId - one service has multiple forms?

ingmarvali replied:
One service has multiple versions. Each version has new eFormId. I was also thinking of one service having multiple forms, but I thought to make the first version with one form.

sleidig commented on api/GovStack_Registration_BB_API_applicant.json:
```diff
@@ -0,0 +1,982 @@
+		"/api/task": {
+			"get": {
+				"tags": ["Operators services"],
+				"summary": "Retrieve all tasks",
```
What is a task in our context? Is this something an operator has to look at?

ingmarvali replied:
Yes, task is for operator(or sometimes for applicant) to look at. One application file is usually being prossesed by one operator. However, sometimes the process flow can be in parallel and be in processing in multiple operators simultaniously.

ingmarvali replied:
task= work that operator must process.  However I just realized, that in many cases the method of assigning tasks is open pool, thus tasks are assigned to a role not to an operator. Should we have additional element "Role"?

ingmarvali committed:
Update ExamplePayload.json

comment added

ingmarvali committed:
Update api/readme.txt

ok

Co-authored-by: Sebastian <sebastian.leidig@gmail.com>

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

update

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

task change

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

ingmarvali committed:
Update GovStack_Registration_BB_API_applicant.json

Upload APi changed, file ID removed

sleidig committed:
merge openapi files into one for easier reference and maintenance

ingmarvali committed:
Merge pull request #10 from sleidig/ingmarvali-API-1_sebastian

merge openapi files into one for easier reference and maintenance

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

ingmarvali committed:
Update GovStack_Registration_BB_API.json

Small adjustments. roleID

ingmarvali committed:
Update OperatorSequenceDiagram.mermaid

ingmarvali committed:
Update GovStack_Registration_BB_API.json

small additions to task services.

ingmarvali committed:
Update ApplicantSequenceDiagram.mermaid

added a task status service

ingmarvali committed:
Update OperatorSequenceDiagram.mermaid

added one document upload api

ingmarvali committed:
Update GovStack_Registration_BB_API.json

ingmarvali commented:
New API services to create application file in Registration BB and process the tasks by operators.
